acceptancetests/assess_persistent_storage.py

```diff
@@ -1,139 +1,428 @@
+
+
+def assert_storage_number(storage_list, expected):
+    log.info('Checking total number of storage unit(s)...')
+    found = len(storage_list)
+    if found != expected:
+        raise JujuAssertionError(
+            'Unexpected number of storage unit(s). '\
+            'Found: {}\nExpected: {}\n.'.format(
+            str(found), str(expected)))
+    else:
+        log.info(
+        'Found {} storage unit(s). Expected: {}.'.format(
+        str(found), str(expected)))
+
+
+def assert_single_blk_existence(storage_list, storage_id):
+    log.info(
+        'Checking existence of single block device storage...'.format(storage_id))
+    if storage_id not in storage_list:
+        raise JujuAssertionError(
+            '{} missing from storage list.'.format(storage_id))
+    else:
+        log.info('{} found in storage list'.format(storage_id))
+
+
+def assert_single_blk_removal(storage_list, storage_id):
+    log.info(
+        'Checking removal of single block device storage...'.format(storage_id))
+    if storage_id in storage_list:
+        raise JujuAssertionError(
+            '{} still exists in storage list.'.format(storage_id))
+    else:
+        log.info('{} has been removed from storage list'.format(storage_id))
+
+
+def assert_persistent_setting(storage_id, found, expected):
+    log.info(
+        'Checking persistent setting of storage unit {}...'.format(
+        storage_id))
+    if found != expected:
+        raise JujuAssertionError(
+            'Incorrect value of persistent setting on storage unit {}. '\
+            'Found: {};\nExpected: {}.'.format(storage_id, found, expected))
+    else:
+        log.info(
+            'Persistent setting of storage unit {}, '\
+            'Found: {}; Expected: {}.'.format(storage_id, found, expected))
+
+
+def assert_storage_status(found_id, expected_id, found_status, expected_status):
+    log.info(
+        'Checking the status of storage {} in volumes...'.format(expected_id))
+    if found_id != expected_id:
+        raise JujuAssertionError(
+            '{} missing from volumes.'.format(expected_id))
+    elif found_status != expected_status:
+        raise JujuAssertionError(
+            'Incorrect status for {}. '\
+            'Found: {}\nExpected: {}'.format(
+            found_id, found_status, expected_status))
+    else:
+        log.info(
+            'The current status of storage {} in volumes is: {}; '\
+            'Expected: {}.'.format(found_id, found_status, expected_status))
+
+def assert_app_removal_msg(client, charm_name):
+    # Run juju remove-application dummy-storage
+    # Bug 1704105: https://bugs.launchpad.net/juju/+bug/1704105
+    # merge_stderr=True is required
+    app_removal_output = client.get_juju_output(
+        'remove-application', charm_name, '--show-log', include_e=False, merge_stderr=True)
+    # pre-set the result to FAILED
+    remove_app_output_check = 'FAILED'
+    for line in app_removal_output.split('\n'):
```

Comment:
Wait, why are you looping through if you break out as soon as you get a match?

This should just be:
```python
if 'will remove unit {}'.format(charm_name) not in app_removal_output:
    raise JujuAssertionError('Missing application name in remove-application output message.')
```